File: device/cuda/reduction/native/reduction.c
#include "reduction.h"
#include "../internal/bridge/core_private.h"

#include <stdio.h>

static const char* g_cuda_reduction_module_source = NULL;

void cuda_reduction_register_module_source(const char* source) {
    g_cuda_reduction_module_source = source;
}

const char* cuda_reduction_module_source(void) {
    return g_cuda_reduction_module_source;
}

void cuda_reduction_status_clear(CUDAStatus* status) {
    cuda_status_clear(status);
}

void cuda_reduction_status_set(CUDAStatus* status, int code, const char* message) {
    cuda_status_set(status, code, message);
}

int cuda_reduction_kernel_name(
    char* out,
    size_t outBytes,
    const char* phase,
    int elementDType,
    CUDAStatus* status
) {
    const char* suffix = cuda_element_dtype_suffix(elementDType);

    if (suffix == NULL) {
        cuda_reduction_status_set(status, -6, "unknown CUDA reduction dtype");
        return -6;
    }

    char prefix[128];
    int written = snprintf(prefix, sizeof(prefix), "reduction_%s", phase);

    if (written <= 0 || (size_t)written >= sizeof(prefix)) {
        cuda_reduction_status_set(status, -6, "CUDA reduction kernel name overflow");
        return -6;
    }

    return cuda_compose_kernel_name(out, outBytes, prefix, suffix, status);
}

File: device/metal/reduction/native/reduction.m
#include "reduction.h"

#include "../internal/bridge/core_private.h"

#include <stdio.h>

void metal_reduction_status_clear(MetalStatus* status) {
    if (status == NULL) {
        return;
    }

    status->code = 0;
    status->message[0] = '\0';
}

void metal_reduction_status_set(MetalStatus* status, int code, const char* message) {
    if (status == NULL) {
        return;
    }

    status->code = code;

    if (message == NULL) {
        status->message[0] = '\0';
        return;
    }

    snprintf(status->message, METAL_STATUS_MESSAGE_BYTES, "%s", message);
}

static const char* metal_reduction_dtype_suffix(int elementDType) {
    switch (elementDType) {
    case MetalElementDTypeFloat32: return "float32";
    case MetalElementDTypeFloat16: return "float16";
    case MetalElementDTypeBFloat16: return "bfloat16";
    default: return NULL;
    }
}

int metal_reduction_kernel_name(
    char* out,
    size_t outBytes,
    const char* phase,
    int elementDType,
    MetalStatus* status
) {
    const char* suffix = metal_reduction_dtype_suffix(elementDType);

    if (phase == NULL || suffix == NULL) {
        metal_reduction_status_set(status, -6, "unknown Metal reduction kernel");
        return -6;
    }

    int written = snprintf(out, outBytes, "reduction_%s_%s", phase, suffix);

    if (written <= 0 || (size_t)written >= outBytes) {
        metal_reduction_status_set(status, -6, "Metal reduction kernel name overflow");
        return -6;
    }

    return 0;
}
